Input encoding cleanup

hrydgard opened the pull request:
Intentionally use u16string/char16_t where characters should be 16-bit.

Might reduce character set confusion problems on platforms that have a 4-byte wchar_t, such as Android.

Not really tested much yet unfortunately. 

Might help #12732, or at least help debug it.

hrydgard committed:
Text input: Clearly use 16-bit chars like the PSP, don't rely on wchar_t.

hrydgard committed:
Support a few more wstring functions on non-windows platforms, due to the armips assembler.

hrydgard commented:
The ios build keeps failing with "An error occurred while generating the build script", that can't be good.

unknownbrackets commented:
That's been happening, Travis problems I think...

-[Unknown]

unknownbrackets commented on ext/native/util/text/utf8.cpp:
```diff
@@ -477,9 +477,9 @@ std::string ConvertWStringToUTF8(const std::wstring &wstr) {
+	const char16_t *const destEnd = dest + destSize;
 
 	UTF8 utf(source.c_str());
 
```
We shouldn't check `sizeof(wchar_t) here anymore, then, right?

-[Unknown]

hrydgard replied:
Indeed, oops.

unknownbrackets replied:
Oops, my bad.

-[Unknown]

unknownbrackets commented on ext/native/util/text/utf8.cpp:
```diff
@@ -477,12 +477,69 @@ std::string ConvertUCS16ToUTF8(const std::u16string &wstr) {
+	char16_t *destw = (char16_t *)dest;
+	const char16_t *const destwEnd = destw + destSize;
+
+	// TODO: Instead of encoding UTF16 here we should encode UCS2.
```
Note: they are the same as long as it's within BMP.  You can encode UCS-2 by encoding to UTF-16 and failing/ignoring any 16-bit unit produced by UTF-16 in the range U+D800 — U+DFFF.

-[Unknown]

hrydgard replied:
Right, I'll change the comment.

unknownbrackets commented on Core/Debugger/SymbolMap.cpp:
```diff
@@ -1003,13 +1003,15 @@ void SymbolMap::GetLabels(std::vector<LabelDefinition> &dest) {
 	if (activeNeedUpdate_)
 		UpdateActiveSymbols();
 
+#ifdef _WIN32
```
Do we have to?  This is not Windows only, and is used by the web debugger.  It should work on all platforms.

-[Unknown]

hrydgard replied:
Whoops, I first added that in an attempt to get it to compile, then went ahead and implemented for the other platforms. Will remove.

hrydgard committed:
Address feedback

hrydgard commented:
Alright, addressed the comments (except for ignoring non-UCS2 characters, left that as TODO). Though come to think of it, since we allow "native" input, should probably take care of that.... so people don't break things by entering emoji, or whatever.

unknownbrackets reviewed (approved):
Right.  But still, should be strictly better than before.... and may hopefully fix or at least better expose that weird issue.

-[Unknown]

hrydgard committed:
More correctly handle UCS2 encoding.

hrydgard commented:
That should do, I think.

unknownbrackets commented:
Still have some uint16_t...

-[Unknown]

hrydgard committed:
Remove some more uint16_t
